File: php/app/service/MolizhishuClient.php
<?php
declare(strict_types=1);

namespace app\service;

use app\exception\MolizhishuApiException;
use think\facade\Config;
use think\facade\Log;

class MolizhishuClient
{
    private string $baseUrl;
    private string $cityUrl;
    private int $timeout;

    public function __construct(private readonly ?EnvFileService $env = null)
    {
        $this->baseUrl = rtrim((string) Config::get('molizhishu.base_url'), '/');
        $this->cityUrl = (string) Config::get('molizhishu.city_url');
        $this->timeout = (int) Config::get('molizhishu.timeout', 30);
    }

    public function submitTask(array $payload, string $source = 'local-api:submit-task'): array
    {
        return $this->request('POST', '/task/batch/shared', $payload, $source);
    }

    public function getCallbackUrl(string $source = 'local-api:callback-url:get'): mixed
    {
        return $this->request('GET', '/task/callback-url', null, $source);
    }

    public function updateCallbackUrl(?string $callbackUrl, string $source = 'local-api:callback-url:update'): mixed
    {
        return $this->request('PUT', '/task/callback-url', ['callbackUrl' => $callbackUrl], $source);
    }

    public function getTaskStatus(string $taskId, string $source = 'local-api:manual-compensation:status'): array
    {
        return $this->request('GET', '/task/status/' . rawurlencode($taskId), null, $source);
    }

    public function getTaskResult(string $taskId, string $source = 'local-api:manual-compensation:result'): array
    {
        return $this->request('GET', '/task/result/' . rawurlencode($taskId), null, $source);
    }

    public function stopTask(string $taskId, string $source = 'local-api:stop-task'): mixed
    {
        return $this->request('PUT', '/task/' . rawurlencode($taskId) . '/stop', null, $source);
    }

    public function getCities(string $source = 'local-api:cities'): array
    {
        return $this->request('GET', $this->cityUrl, null, $source, true);
    }

    private function request(string $method, string $pathOrUrl, ?array $payload, string $source, bool $absolute = false): mixed
    {
        $token = $this->env?->get('MOLIZHISHU_TOKEN') ?: (string) Config::get('molizhishu.token', '');
        if ($token === '') {
            throw new MolizhishuApiException('MOLIZHISHU_TOKEN 未配置');
        }

        $url = $absolute ? $pathOrUrl : $this->baseUrl . $pathOrUrl;
        $started = microtime(true);
        $headers = [
            'Accept: application/json',
            'Authorization: Bearer ' . $token,
        ];

        $ch = curl_init($url);
        curl_setopt_array($ch, [
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_CUSTOMREQUEST => $method,
            CURLOPT_TIMEOUT => $this->timeout,
            CURLOPT_HTTPHEADER => $headers,
        ]);

        if ($payload !== null) {
            $body = json_encode($payload, JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES);
            curl_setopt($ch, CURLOPT_POSTFIELDS, $body);
            curl_setopt($ch, CURLOPT_HTTPHEADER, [...$headers, 'Content-Type: application/json']);
        }

        $body = curl_exec($ch);
        $curlError = curl_error($ch);
        $httpStatus = (int) curl_getinfo($ch, CURLINFO_RESPONSE_CODE);
        if (PHP_VERSION_ID < 80500) {
            curl_close($ch);
        }

        $duration = (int) round((microtime(true) - $started) * 1000);

        if ($body === false) {
            Log::error(sprintf('[molizhishu] source=%s method=%s url=%s http_status=0 success=false code=null message="%s" duration=%dms', $source, $method, $url, $curlError, $duration));
            throw new MolizhishuApiException('模力指数接口网络异常：' . $curlError);
        }

        $decoded = json_decode((string) $body, true);
        if (!is_array($decoded)) {
            Log::error(sprintf('[molizhishu] source=%s method=%s url=%s http_status=%d success=false code=null message="invalid-json" duration=%dms', $source, $method, $url, $httpStatus, $duration));
            throw new MolizhishuApiException('模力指数接口返回非 JSON', null, $httpStatus, (string) $body);
        }

        $success = (bool) ($decoded['success'] ?? false);
        $code = isset($decoded['code']) ? (int) $decoded['code'] : null;
        $message = (string) ($decoded['message'] ?? '');
        Log::info(sprintf('[molizhishu] source=%s method=%s url=%s http_status=%d success=%s code=%s message="%s" duration=%dms', $source, $method, $url, $httpStatus, $success ? 'true' : 'false', $code === null ? 'null' : (string) $code, addslashes($message), $duration));

        if ($httpStatus < 200 || $httpStatus >= 300) {
            throw new MolizhishuApiException('模力指数 HTTP 异常：' . $httpStatus, $code, $httpStatus, (string) $body);
        }

        if (!$success) {
            throw new MolizhishuApiException($message !== '' ? $message : '模力指数业务处理失败', $code, $httpStatus, (string) $body);
        }

        return $decoded['data'] ?? null;
    }
}

File: python/app/molizhishu_client.py
import logging
import time
from typing import Any

import httpx

from .config import Settings

logger = logging.getLogger(__name__)


class MolizhishuApiError(Exception):
    def __init__(self, message: str, code: int | None = None, http_status: int = 502):
        super().__init__(message)
        self.code = code
        self.http_status = http_status


class MolizhishuClient:
    def __init__(self, settings: Settings):
        self.settings = settings

    async def submit_task(self, payload: dict[str, Any]) -> dict[str, Any]:
        return await self._request("POST", "/task/batch/shared", payload, "local-api:submit-task")

    async def get_task_status(self, task_id: str, source: str) -> dict[str, Any]:
        return await self._request("GET", f"/task/status/{task_id}", None, source)

    async def get_task_result(self, task_id: str, source: str) -> dict[str, Any]:
        return await self._request("GET", f"/task/result/{task_id}", None, source)

    async def stop_task(self, task_id: str) -> Any:
        return await self._request("PUT", f"/task/{task_id}/stop", None, "local-api:stop-task")

    async def get_callback_url(self) -> Any:
        return await self._request("GET", "/task/callback-url", None, "local-api:callback-url:get")

    async def update_callback_url(self, callback_url: str | None) -> Any:
        return await self._request("PUT", "/task/callback-url", {"callbackUrl": callback_url}, "local-api:callback-url:update")

    async def get_cities(self) -> Any:
        return await self._request("GET", self.settings.city_url, None, "local-api:cities", absolute=True)

    async def _request(self, method: str, path_or_url: str, payload: dict[str, Any] | None, source: str, absolute: bool = False) -> Any:
        if not self.settings.token:
            raise MolizhishuApiError("MOLIZHISHU_TOKEN 未配置", None, 0)

        url = path_or_url if absolute else self.settings.base_url.rstrip("/") + path_or_url
        started = time.perf_counter()
        headers = {"Authorization": f"Bearer {self.settings.token}", "Accept": "application/json"}
        try:
            async with httpx.AsyncClient(timeout=self.settings.timeout_seconds) as client:
                response = await client.request(method, url, json=payload, headers=headers)
        except httpx.HTTPError as exc:
            logger.exception("[molizhishu] source=%s method=%s url=%s http_status=0 success=false", source, method, url)
            raise MolizhishuApiError(f"模力指数接口网络异常：{exc}") from exc

        try:
            envelope = response.json()
        except ValueError as exc:
            raise MolizhishuApiError("模力指数接口返回非 JSON", None, response.status_code) from exc

        success = bool(envelope.get("success"))
        code = envelope.get("code")
        message = envelope.get("message") or ""
        logger.info(
            "[molizhishu] source=%s method=%s url=%s http_status=%s success=%s code=%s message=%r duration=%sms",
            source, method, url, response.status_code, success, code, message, round((time.perf_counter() - started) * 1000),
        )

        if response.status_code < 200 or response.status_code >= 300:
            raise MolizhishuApiError(f"模力指数 HTTP 异常：{response.status_code}", code, response.status_code)
        if not success:
            raise MolizhishuApiError(message or "模力指数业务处理失败", code, 502)
        return envelope.get("data")
